ducktape/tests/runner_client.py

```
@@ -423,14 +429,30 @@ def setup_test(self):
         self.log(logging.INFO, "Setting up...")
         self.test.setup()
 
-    def run_test(self):
+    def run_test(self, test_runner_timeout):
         """Run the test!
 
         We expect test_context.function to be a function or unbound method which takes an
         instantiated test object as its argument.
         """
+        timeout_seconds = int(test_runner_timeout / 1000)
+
+        def _handle_timeout(signum, frame):
+            raise TestTimeoutError(f"Test timed out after {timeout_seconds} seconds")
+
         self.log(logging.INFO, "Running...")
-        return self.test_context.function(self.test)
+
+        # Set up signal alarm
+        signal.signal(signal.SIGALRM, _handle_timeout)
```

So how does this actually work? An exception will just be thrown from the top of whatever stack running at the moment this occurs? That's kind of neat (you can't do that in C++ from a signal).

Are `except` blocks active in the current stack executed? Doesn't this mean the test could just end up swallowing the exception? OTOH it at least means that cleanup is more likely to happen.

---

Hmm, so learned about throwing from signal handler, which does work in python (it's how keyboard interrupt works too).

I feel like there is a high chance of the exception being swallowed, e.g,. because the running test is using `wait_until` which can swallow exceptions.

---

> An exception will just be thrown from the top of whatever stack running at the moment this occurs?

that's right!! currently, timeouts have no indication of what went wrong with the test. now we'll be able to diagnose better

> I feel like there is a high chance of the exception being swallowed, e.g,. because the running test is using `wait_until` which can swallow exceptions.

yeah, the timeout should exceed the maximum duration expected for any test

---

What I mean is imagine a test is "hanging" because it's just doing too much stuff, and it takes a long time, and it hits the test timeout: we send a signal but this just gets eaten by a 

```
except Exception as e:
    pass # or something else, but do not rethrow
```

Then the test wlil just carry on, and the whole run will continue on beyond the `--test-runner-timeout` because the server/runner no longer enforces it. Right?

---

Oh I see.. you're right! let me check if it is somehow handled in `pytest` that uses signals to handle timeouts

---

Just for my understanding, ducktape relies on the services running on nodes to clean themselves up, right? And this happens in test teardown.

So if we force kill a process, either as the first approach (replacing this one), or as a second approach (i.e., we try the signal first, but if it fails, we force kill the process) it will leave behind running services, files on disk and stuff like that, right? So it may not be safe to continue the test run since these nodes are in an unclean state?

---

added another error type for cases when the exception is suppressed by the test `except` block: https://github.com/redpanda-data/ducktape/compare/32b824514933d40ecb263d245bc0acdf58559ed0..0ab739f57b5cb7c3bf65c2a5316e98720f6f46ce

> Just for my understanding, ducktape relies on the services running on nodes to clean themselves up, right?

right, tearDown will still run even on the new exception introduced here.

> So if we force kill a process, either as the first approach (replacing this one), or as a second approach (i.e., we try the signal first, but if it fails, we force kill the process) it will leave behind running services, files on disk and stuff like that, right? So it may not be safe to continue the test run since these nodes are in an unclean state?

if the signal succeeds, then it will be handled as a normal exception by the test..
if the signal fails, then the test will indeed be running consuming resources until (1) the test exits OR (2) the timeout on the server hits, and the entire session is dropped without even triggering new tests or waiting for the running ones. what we could do is make the server timeout to `test_runner_timeout + 10 seconds` so that if the signal fails, then the server will raise a "hard" error (after 10 seconds) and ducktape will immediately exit (non-zero code). does that make sense?